File: mcp-clients/sse/python/uv-simple-math-client/math_client.py
#!/usr/bin/env -S uv run
# /// script
# requires-python = ">=3.9"
# dependencies = [
#     "mcp",
# ]
# ///

import asyncio
import sys
from typing import Optional
from mcp import ClientSession
from mcp.client.sse import sse_client

class SimpleMathClient:
    def __init__(self):
        self.session: Optional[ClientSession] = None
        self._streams_context = None
        self._session_context = None

    async def connect_to_server(self, server_url: str):
        """Connect to the math server running with SSE transport"""
        print(f"Connecting to server at {server_url}...")
        
        # Create and store the context managers
        self._streams_context = sse_client(url=server_url)
        streams = await self._streams_context.__aenter__()

        self._session_context = ClientSession(*streams)
        self.session = await self._session_context.__aenter__()

        # Initialize the session
        await self.session.initialize()

        # List available tools to verify connection
        response = await self.session.list_tools()
        tools = response.tools
        print("\nConnected to server with tools:", [tool.name for tool in tools])
        
        for tool in tools:
            print(f"  - {tool.name}: {tool.description}")

    async def cleanup(self):
        """Properly clean up the session and streams"""
        if self._session_context:
            await self._session_context.__aexit__(None, None, None)
        if self._streams_context:
            await self._streams_context.__aexit__(None, None, None)

    async def call_math_tool(self, tool_name: str, a: float, b: float):
        """Call a math tool on the server"""
        if not self.session:
            print("Error: Not connected to server")
            return None
        
        try:
            result = await self.session.call_tool(tool_name, {"a": a, "b": b})
            return result.content
        except Exception as e:
            print(f"Error calling {tool_name}: {str(e)}")
            return None

    async def chat_loop(self):
        """Run an interactive chat loop for math operations"""
        print("\nSimple Math Client Started!")
        print("Format: <operation> <number1> <number2>")
        print("Available operations: add, subtract, multiply, divide")
        print("Example: add 5 3")
        print("Type 'quit' to exit.")
        
        while True:
            try:
                command = input("\n> ").strip()
                
                if command.lower() == 'quit':
                    break
                
                parts = command.split()
                if len(parts) != 3:
                    print("Invalid format. Use: <operation> <number1> <number2>")
                    continue
                
                operation, a_str, b_str = parts
                
                try:
                    a = float(a_str)
                    b = float(b_str)
                except ValueError:
                    print("Numbers must be valid integers or floats")
                    continue
                
                if operation not in ['add', 'subtract', 'multiply', 'divide']:
                    print("Unknown operation. Available: add, subtract, multiply, divide")
                    continue
                
                result = await self.call_math_tool(operation, a, b)
                if result is not None:
                    print(f"Result: {result}")
                    
            except Exception as e:
                print(f"Error: {str(e)}")

async def main():
    if len(sys.argv) < 2:
        print("Usage: uv run math_client.py <URL of SSE MCP server>")
        print("Example: uv run math_client.py http://localhost:8000/sse")
        sys.exit(1)

    client = SimpleMathClient()
    try:
        await client.connect_to_server(server_url=sys.argv[1])
        await client.chat_loop()
    finally:
        await client.cleanup()

if __name__ == "__main__":
    asyncio.run(main())

File: mcp-clients/sse/typescript/math-client-w-progress/src/math_client.ts
#!/usr/bin/env node

/**
 * Math Client Example using MCP over SSE
 * 
 * This script demonstrates a simple MCP client that connects to a math server using SSE and calls math tools.
 */

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import { createInterface } from 'readline/promises';
import { URL } from 'url';
import {
  CallToolResult,
  NotificationSchema
} from "@modelcontextprotocol/sdk/types.js";
import { z } from "zod";

// Get server URL from command line argument
const serverUrl = process.argv[2];
if (!serverUrl) {
  console.error("Please provide the server URL as a command line argument");
  console.error("Example: npm run dev http://localhost:8000/sse");
  process.exit(1);
}

async function main() {
  const url = new URL(serverUrl);
  const transport = new SSEClientTransport(url);
  
  // Create MCP client
  const client = new Client({ name: "Math Client", version: "1.0.0" });
  
  // Define a custom notification schema
  const CustomNotificationSchema = NotificationSchema.extend({
    method: z.literal("notifications/custom"),
    params: z.object({
      operation: z.string(),
      status: z.string(),
      message: z.string(),
      sessionId: z.string().optional()
    })
  });
  
  try {
    // Connect to the server
    console.log(`Connecting to math server at ${serverUrl}...`);
    await client.connect(transport);
    console.log("Connected successfully!");
    
    // Register a notification handler for custom notifications
    client.setNotificationHandler(CustomNotificationSchema, (notification) => {
      console.log(`📢 NOTIFICATION: ${notification.params.message} (${notification.params.status})`);
    });
    
    // Get server information
    const serverInfo = client.getServerVersion();
    console.log(`Connected to: ${serverInfo?.name} v${serverInfo?.version}`);
    
    // List available tools
    const tools = await client.listTools();
    console.log("Available tools:");
    tools.tools.forEach(tool => {
      console.log(`- ${tool.name}: ${tool.description}`);
    });
    
    // Setup readline interface for user input
    const readline = createInterface({
      input: process.stdin,
      output: process.stdout
    });
    
    console.log("\nMath Client Ready. Type 'exit' to quit.");
    console.log("Example: add 5 3");
    
    while (true) {
      const input = await readline.question("\nEnter operation (add/subtract/multiply/divide) and numbers: ");
      
      if (input.toLowerCase() === 'exit') {
        break;
      }
      
      const parts = input.trim().split(/\s+/);
      if (parts.length < 3) {
        console.log("Invalid input. Please use format: operation number1 number2");
        continue;
      }
      
      const operation = parts[0].toLowerCase();
      const num1 = parseFloat(parts[1]);
      const num2 = parseFloat(parts[2]);
      
      if (isNaN(num1) || isNaN(num2)) {
        console.log("Invalid numbers. Please enter valid numbers.");
        continue;
      }

      console.log(`Calling ${operation} with ${num1} and ${num2}`);
      
      try {
        // Display a waiting message
        console.log(`Waiting for result from server...`);
        
        // Start a timer to show that we're waiting
        let waitingDots = '';
        const waitingInterval = setInterval(() => {
          waitingDots = waitingDots.length < 3 ? waitingDots + '.' : '';
          process.stdout.write(`\rStill waiting${waitingDots}   `);
        }, 500);
        
        // Call the tool
        const result = await client.callTool({
          name: operation,
          arguments: { a: num1, b: num2 }
        }) as CallToolResult;
        
        // Clear the waiting indicator
        clearInterval(waitingInterval);
        process.stdout.write('\r                      \r');
        
        console.log(`Result: ${result.content?.[0]?.text}`);
      } catch (error) {
        console.error(`Error calling ${operation}:`, error);
      }
    }
    
    readline.close();
    
  } catch (error) {
    console.error("Error:", error);
  } finally {
    // Close the client connection
    await client.close();
    console.log("Connection closed");
  }
}

main().catch(error => {
  console.error("Unhandled error:", error);
  process.exit(1);
});